"""
Author: Susheel Bhanu BUSI
Affiliation: ESB group LCSB UniLU
Date: [2021-01-31]
Run: snakemake -s Snakefile --use-conda --cores 40 -rp
Latest modification:
"""

import glob

configfile:"config.yaml"
DATA_DIR=config['data_dir']
RESULTS_DIR=config['results_dir']
SAMPLES=[line.strip() for line in open("prokaryote_complete_accessions.txt", 'r')]    # if using a sample list instead of putting them in a config file

###########
rule all:
    input:
        expand(os.path.join(RESULTS_DIR, "{sample}"), sample=SAMPLES)

############################
# Download Genbank genomes #
############################
rule coverage:
    output:
        directory(os.path.join(RESULTS_DIR, "{sample}"))
    log:
        os.path.join("logs/download.{sample}.log")
    conda:
        os.path.join("envs/ncbi-genome-download.yaml")
    message:
        "Downloading FASTA, GFF and Protein-FASTA for {wildcards.sample}"
    shell:
        "(date && ncbi-genome-download -A {wildcards.sample} all -s genbank --formats fasta,gff,protein-fasta --no-cache --verbose && date) &> {log}"

